pwnlib/elf/pltgot.py

```diff
@@ -0,0 +1,291 @@
+from pwnlib.util import packing
+from pwnlib.log import getLogger
+
+log = getLogger(__name__)
+
+def read_got(elf):
+    """ Read GOT symbols and corresponding program locations from the binary. """
+
+    got = {}
+
+    # TODO: In reality, the canonical relocation list is the one linked from
+    # the DYNAMIC segment; if we parse that, then we can get GOT even if the
+    # section headers are missing or obfuscated away.
+    for sect in elf.iter_sections():
+        if not isinstance(sect, RelocationSection):
+            continue
+
+        if sect.header.sh_link == SHN_INDICES.SHN_UNDEF:
+            continue
+
+        # We have to examine all relocation sections.
+        # Binaries (especially RELRO binaries) may put PLT GOT entries in *either*
+        # .rel[a].dyn or .rel[a].plt.
+        # This might pull in extraneous relocations, but it's better than missing
+        # important external symbols.
+
+        # Find the symbols for the relocation section
+        sym_rel = elf.get_section(sect.header.sh_link)
+
+        # Populate the GOT
+        for rel in sect.iter_relocations():
+            sym_idx  = rel.entry.r_info_sym
+            symbol   = sym_rel.get_symbol(sym_idx)
+            if not symbol:
+                # local relocation
+                continue
+            name = symbol.name
+
+            got[name] = rel.entry.r_offset
+
+    return got
+
+
+def _parse_plt_i386(symbols, revgot, pltsect):
+    res = {}
+    data = pltsect.data()
+    saddr = pltsect.header.sh_addr
+    pos = 0
+
+    # Need the got base address for PIE executables
+    # TODO: how might we get this if the symbol table is missing?
+    gotaddr = symbols.get('_GLOBAL_OFFSET_TABLE_', 0)
+
+    while pos < len(data):
+        if data[pos:pos+2] == '\xff\x25':
+            # jmp dword ptr ds:ABS (non-PIE)
+            addr = packing.u32(data[pos+2:pos+6])
+            if addr in revgot:
+                res[revgot[addr]] = saddr + pos
+            pos += 8
+        elif data[pos:pos+2] == '\xff\xa3':
+            # jmp dword ptr [ebx+REL] (PIE)
+            addr = packing.u32(data[pos+2:pos+6]) + gotaddr
+            if addr in revgot:
+                res[revgot[addr]] = saddr + pos
+            pos += 8
+        else:
+            # unknown/stub (add more cases for future/different PLT stub designs)
+            pos += 4
+
+    return res
+
+def _parse_plt_amd64(symbols, revgot, pltsect):
+    res = {}
+    data = pltsect.data()
+    saddr = pltsect.header.sh_addr
+    pos = 0
+
+    while pos < len(data):
+        if data[pos:pos+2] == '\xff\x25':
+            # jmp dword ptr [rip+REL]
+            addr = packing.u32(data[pos+2:pos+6]) + saddr + pos + 6
+            if addr in revgot:
+                res[revgot[addr]] = saddr + pos
+            pos += 8
+        else:
+            # unknown/stub (add more cases for future/different PLT stub designs)
+            pos += 4
+
+    return res
+
+def _parse_plt_arm(symbols, revgot, pltsect):
```

Comment:
We already rely on Capstone vis-a-vis `ROPgadget`.  This code should be updated to extract the values via `capstone`'s disassembler.